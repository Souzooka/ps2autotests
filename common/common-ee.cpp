#include <stdio.h>
#include <stdarg.h>
#include <string.h>
#include <kernel.h>

#define TEST_THREAD_PRIORITY   0x40
#define TEST_THREAD_STACK_SIZE 0x8000

char schedfBuffer[65536];
unsigned int schedfBufferPos = 0;

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	schedfBufferPos += vsprintf(schedfBuffer + schedfBufferPos, format, args);
	va_end(args);
}

void flushschedf() {
	printf("%s", schedfBuffer);
	schedfBuffer[0] = '\0';
	schedfBufferPos = 0;
}

u32 g_testThreadDoneSema = 0;

extern int test_main(int argc, char *argv[]);

char testThreadStack[TEST_THREAD_STACK_SIZE] __attribute__ ((aligned(16)));
void testThreadProc() {
	//TODO: Print begin/end traces
	//TODO: Pass arguments from main thread to the test function
	test_main(0, NULL);
	SignalSema(g_testThreadDoneSema);
}

int main(int argc, char *argv[]) {
	//A thread is created to ensure the execution environment is the same
	//across all possible boot methods (direct hw boot, ps2link, emulator, etc.)
	
	ee_sema_t semaInfo;
	memset(&semaInfo, 0, sizeof(ee_sema_t));
	semaInfo.init_count = 0;
	semaInfo.max_count  = 1;
	semaInfo.option     = 0;
	g_testThreadDoneSema = CreateSema(&semaInfo);
	
	u32 testThreadId = 0;
	ee_thread_t threadParam;
	memset(&threadParam, 0, sizeof(ee_thread_t));
	threadParam.func               = (void*)&testThreadProc;
	threadParam.initial_priority   = TEST_THREAD_PRIORITY;
	threadParam.stack_size         = TEST_THREAD_STACK_SIZE;
	threadParam.stack              = testThreadStack;
	testThreadId = CreateThread(&threadParam);
	
	StartThread(testThreadId, NULL);
	WaitSema(g_testThreadDoneSema);
	
	return 0;
}
