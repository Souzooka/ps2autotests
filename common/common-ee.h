#ifndef PS2AUTOTESTS_COMMON_EE_H
#define PS2AUTOTESTS_COMMON_EE_H

// Defines for MSVC highlighting.  Not intended to actually compile with msc.
#ifdef _MSC_VER
#define __STDC__
#define _EE
#define __attribute__(x)
#endif

#include <stdio.h>
#include <tamtypes.h>

#undef main
#define main test_main

void schedf(const char *format, ...);
void flushschedf();

#endif
