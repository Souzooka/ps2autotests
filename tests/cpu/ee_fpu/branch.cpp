#include <common-ee.h>

enum BranchResultFlags {
	BRANCH_SKIPPED = 0,
	BRANCH_FOLLOWED = 1,
	BRANCH_SET_RA = 2,
	BRANCH_DELAY_SLOT = 4,
};

#define BO_OP_FUNC(OP) \
static inline void BO_##OP(u32 &result) { \
	asm volatile ( \
		".set noreorder\n" \
		"" \
		"lui %0, 0\n" \
		"or $t9, $0, $ra\n" \
		"" \
		#OP " " #OP "_branch%=\n" \
		"ori %0, %0, 4\n" \
		"" \
		"b " #OP "_skip%=\n" \
		"nop\n" \
		"" \
		#OP "_branch%=:\n" \
		"ori %0, %0, 1\n" \
		"" \
		#OP "_skip%=:\n" \
		"beq $t9, $ra, " #OP "_done%=\n" \
		"nop\n" \
		"" \
		"or $ra, $0, $t9\n" \
		"ori %0, %0, 2\n" \
		"" \
		#OP "_done%=:\n" \
		"" \
		: "=&r"(result) : : "t9" \
	); \
}

static inline void PRINT_BRANCH(const u32 result, bool newline) {
	printf("%s, ", (result & BRANCH_FOLLOWED) != 0 ? "followed" : "skipped");
	printf("%s, ", (result & BRANCH_SET_RA) != 0 ? "set ra" : "ignored ra");
	printf("%s", (result & BRANCH_DELAY_SLOT) != 0 ? "ran delay slot" : "skipped delay slot");
	if (newline) {
		printf("\n");
	}
}

static inline void SET_FC_I(int flag) {
	register u32 mask = ~(1 << 23);
	register u32 bit = flag << 23;

	asm volatile (
		"cfc1 $t6, $31\n"
		"and $t6, $t6, %0\n"
		"or $t6, $t6, %1\n"
		"ctc1 $t6, $31\n"
		: : "r"(mask), "r"(bit) : "t6"
	);
}

static inline void SET_FC_CMP_TRUE() {
	asm volatile (
		"c.eq.s $f0, $f0\n"
	);
}

static inline void SET_FC_CMP_FALSE() {
	asm volatile (
		"c.f.s $f0, $f0\n"
	);
}

#define TEST_BO_FUNC(OP) \
BO_OP_FUNC(OP); \
static void test_##OP() { \
	u32 res; \
	 \
	printf("%s:\n", #OP); \
	 \
	SET_FC_I(1); \
	BO_##OP(res); \
	printf("  %s 1: ", #OP); PRINT_BRANCH(res, true); \
	 \
	SET_FC_I(0); \
	BO_##OP(res); \
	printf("  %s 0: ", #OP); PRINT_BRANCH(res, true); \
	 \
	SET_FC_CMP_TRUE(); \
	BO_##OP(res); \
	printf("  %s c.eq: ", #OP); PRINT_BRANCH(res, true); \
	 \
	SET_FC_CMP_FALSE(); \
	BO_##OP(res); \
	printf("  %s c.f: ", #OP); PRINT_BRANCH(res, true); \
	\
	printf("\n"); \
}

TEST_BO_FUNC(bc1f);
TEST_BO_FUNC(bc1fl);
TEST_BO_FUNC(bc1t);
TEST_BO_FUNC(bc1tl);

int main(int argc, char **argv) {
	printf("-- TEST BEGIN\n");

	test_bc1f();
	test_bc1fl();
	test_bc1t();
	test_bc1tl();

	printf("-- TEST END\n");

	return 0;
}
