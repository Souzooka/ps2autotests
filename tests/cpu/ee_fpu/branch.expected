-- TEST BEGIN
bc1f:
  bc1f 1: skipped, ignored ra, ran delay slot
  bc1f 0: followed, ignored ra, ran delay slot
  bc1f c.eq: skipped, ignored ra, ran delay slot
  bc1f c.f: followed, ignored ra, ran delay slot

bc1fl:
  bc1fl 1: skipped, ignored ra, skipped delay slot
  bc1fl 0: followed, ignored ra, ran delay slot
  bc1fl c.eq: skipped, ignored ra, skipped delay slot
  bc1fl c.f: followed, ignored ra, ran delay slot

bc1t:
  bc1t 1: followed, ignored ra, ran delay slot
  bc1t 0: skipped, ignored ra, ran delay slot
  bc1t c.eq: followed, ignored ra, ran delay slot
  bc1t c.f: skipped, ignored ra, ran delay slot

bc1tl:
  bc1tl 1: followed, ignored ra, ran delay slot
  bc1tl 0: skipped, ignored ra, skipped delay slot
  bc1tl c.eq: followed, ignored ra, ran delay slot
  bc1tl c.f: skipped, ignored ra, skipped delay slot

-- TEST END
